### PR TITLE
Attach Detach Controller CRD validations

### LINKED ISSUE #68260: Attach Detach Controller Install CRD does not specify validation schema
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:

The CRD installation in https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/attach_detach_controller.go#L671 does not specify validation.

```GO
func (adc *attachDetachController) installCRDs() error {
	crd := &apiextensionsv1beta1.CustomResourceDefinition{
		ObjectMeta: metav1.ObjectMeta{
			Name: csiapiv1alpha1.CsiDriverResourcePlural + "." + csiapiv1alpha1.GroupName,
		},
		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
			Group:   csiapiv1alpha1.GroupName,
			Version: csiapiv1alpha1.SchemeGroupVersion.Version,
			Scope:   apiextensionsv1beta1.ClusterScoped,
			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
				Plural: csiapiv1alpha1.CsiDriverResourcePlural,
				Kind:   reflect.TypeOf(csiapiv1alpha1.CSIDriver{}).Name(),
			},
		},
	}
```

@liggitt pointed out lack of validation may cause issues with watchers and listers. So we validation should be added.

/milestone v1.12

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by hoegaarden**
**What this PR does / why we need it**:

> @liggitt pointed out lack of validation may cause issues with watchers and listers. So we validation should be added.

**Which issue(s) this PR fixes**:
Fixes #68260

**Special notes for your reviewer**:

We are not sure if those validations are narrow enough -- we don't set any of the fields as mandatory, we do no specify any REs, min/max lengths, ...

**Release note**:
```release-note
Add validations to volume A/D controller CRDs
```

/assign @kubernetes/sig-storage-pr-reviews

### CODE REVIEW

**Comment by hoegaarden** [comment]
@liggitt -- Thanks for your comments!

> The minimum required is that we check the types of all the fields defined in the go types to ensure we can decode into the go types successfully.

I believe we have done that -- but let's check (again) once we've moved the CRD definition into yaml.

> @lavalamp requested the CRD be defined in a yaml file in the original API PR, and I agree with that request. Reviewing and tracking changes to API type definitions made in code like this is difficult and error prone.

OK. Sorry, we didn't see that. Could you point us to that request?

We try to change the PR so that it loads the CRD definition from a file.

> Additionally, that yaml file should be in a package covered by API review and approval

I am not sure what that exactly means. Could you please add more details on what that means for our PR?

Thanks!

/cc @mariantalla

**Comment by guineveresaenger** [comment]
@liggit @hoegaarden - since the related issue is in the 1.12 milestone, can you make sure the labels match? Otherwise this will hang in CI as we're in code freeze.
Specifically, we are missing:
- milestone
- priority

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 1]
```diff
@@ -667,50 +667,114 @@ func (adc *attachDetachController) processVolumesInUse(
...
+
+// installCRDs creates the specified CustomResourceDefinition for the CSIDrivers object.
+func (adc *attachDetachController) installCRDs() error {
+	crd := getCSIDriverCRD()
+
+	res, err := adc.crdClient.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
+
+	if err == nil {
+		glog.Infof("CSIDrivers CRD created successfully: %#v",
+			res)
```
Comment (neolit123) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: please move this on the previous line. [1]

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 2]
```diff
@@ -667,50 +667,114 @@ func (adc *attachDetachController) processVolumesInUse(
...
+	} else if apierrors.IsAlreadyExists(err) {
+		glog.Warningf("CSIDrivers CRD already exists: %#v, err: %#v",
+			res, err)
```
Comment (neolit123) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: [1]
also, just `%v` is fine for `error`

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 3]
```diff
@@ -667,50 +667,114 @@ func (adc *attachDetachController) processVolumesInUse(
...
+	} else {
+		glog.Errorf("failed to create CSIDrivers CRD: %#v, err: %#v",
+			res, err)
```
Comment (neolit123) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: [1]
also, just `%v` is fine for `error`

File: pkg/controller/volume/attachdetach/attach_detach_controller_test.go [hunk 4]
```diff
@@ -306,3 +306,78 @@ func attachDetachRecoveryTestCase(t *testing.T, extraPods1 []*v1.Pod, extraPods2
...
+	if !hasSpec {
+		t.Fatal("Expected to find a validator for 'spec'")
+	}
+
+	attachRequired, hasAttachRequired := spec.Properties["attachRequired"]
+	if !hasAttachRequired {
+		t.Fatal("Expected to find validator for 'spec.attachRequired'")
+	}
+	if attachRequired.Type != "boolean" {
+		t.Fatalf("Expected 'spec.attachRequired' to have a type validator for boolean, got %s", attachRequired.Type)
```
Comment (neolit123) [inline pkg/controller/volume/attachdetach/attach_detach_controller_test.go]: `got` -> `got: ` [2]

File: pkg/controller/volume/attachdetach/attach_detach_controller_test.go [hunk 5]
```diff
@@ -306,3 +306,78 @@ func attachDetachRecoveryTestCase(t *testing.T, extraPods1 []*v1.Pod, extraPods2
...
+	if crd.Spec.Validation == nil {
+		t.Fatal("Expected NodeInfo CRD to have validations configured")
+	}
+
+	drivers, hasDrivers := crd.Spec.Validation.OpenAPIV3Schema.Properties["csiDrivers"]
+	if !hasDrivers {
+		t.Fatal("Expected to find validator for 'csiDrivers'")
+	}
+	if drivers.Type != "array" {
+		t.Fatalf("Expected 'csiDrivers' to have a type validator for array, got %s", drivers.Type)
```
Comment (neolit123) [inline pkg/controller/volume/attachdetach/attach_detach_controller_test.go]: [2]

File: pkg/controller/volume/attachdetach/attach_detach_controller_test.go [hunk 6]
```diff
@@ -306,3 +306,78 @@ func attachDetachRecoveryTestCase(t *testing.T, extraPods1 []*v1.Pod, extraPods2
...
+	}
+	if drivers.Items.Schema == nil {
+		t.Fatal("Expected 'csiDrivers' to specify a schema for array items")
+	}
+
+	schema := drivers.Items.Schema
+
+	driver := schema.Properties["driver"]
+	if driver.Type != "string" {
+		t.Fatalf("Expected 'csiDrivers[].driver' to have a type validator for string, got %s", driver.Type)
```
Comment (neolit123) [inline pkg/controller/volume/attachdetach/attach_detach_controller_test.go]: [2]

File: pkg/controller/volume/attachdetach/attach_detach_controller_test.go [hunk 7]
```diff
@@ -306,3 +306,78 @@ func attachDetachRecoveryTestCase(t *testing.T, extraPods1 []*v1.Pod, extraPods2
...
+	}
+
+	nodeID := schema.Properties["nodeID"]
+	if nodeID.Type != "string" {
+		t.Fatalf("Expected 'csiDrivers[].nodeID' to have a type validator for string, got %s", nodeID.Type)
```
Comment (neolit123) [inline pkg/controller/volume/attachdetach/attach_detach_controller_test.go]: [2]

File: pkg/controller/volume/attachdetach/attach_detach_controller_test.go [hunk 8]
```diff
@@ -306,3 +306,78 @@ func attachDetachRecoveryTestCase(t *testing.T, extraPods1 []*v1.Pod, extraPods2
...
+	}
+
+	topologyKeys := schema.Properties["topologyKeys"]
+	if topologyKeys.Type != "array" {
+		t.Fatalf("Expected 'csiDrivers[].topologyKeys' to have a type validator for array, got %s", topologyKeys.Type)
```
Comment (neolit123) [inline pkg/controller/volume/attachdetach/attach_detach_controller_test.go]: [2]

File: pkg/controller/volume/attachdetach/attach_detach_controller_test.go [hunk 9]
```diff
@@ -306,3 +306,78 @@ func attachDetachRecoveryTestCase(t *testing.T, extraPods1 []*v1.Pod, extraPods2
...
+	}
+	if topologyKeys.Items == nil {
+		t.Fatal("Expected 'csiDrivers[].topologyKeys' to specify properties for array items")
+	}
+	if topologyKeys.Items.Schema == nil {
+		t.Fatal("Expected 'csiDrivers[].topologyKeys' to specify a schema for array items")
+	}
+	topologyKeysSchema := topologyKeys.Items.Schema
+	if topologyKeysSchema.Type != "string" {
+		t.Fatalf("Expected 'csiDrivers[].topologyKeys[]' to have a type validator for string, got %s", topologyKeysSchema.Type)
```
Comment (neolit123) [inline pkg/controller/volume/attachdetach/attach_detach_controller_test.go]: [2]

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 10]
```diff
@@ -667,50 +667,114 @@ func (adc *attachDetachController) processVolumesInUse(
...
+			Validation: &apiextensionsv1beta1.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensionsv1beta1.JSONSchemaProps{
+					Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
+						"spec": apiextensionsv1beta1.JSONSchemaProps{
+							Description: "Specification of the CSI Driver.",
+							Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
+								"attachRequired": apiextensionsv1beta1.JSONSchemaProps{
+									Description: "Indicates this CSI volume driver requires an attach operation," +
+										" and that Kubernetes should call attach and wait for any attach operation to" +
+										" complete before proceeding to mounting.",
```
Comment (neolit123) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: `complete before proceeding to mounting` -> `complete before proceeding to mount`
?

**Review by neolit123 — COMMENTED** [review]
thanks for the PR! @hoegaarden 
added some minor styling and copy edit comments.

**Comment by guineveresaenger** [comment]
/milestone v 1.12
/priority critical-urgent

@hoegaarden please verify and rebase. merging this would make me so happy. :)

**Comment by guineveresaenger** [comment]
/milestone v1.12

(missed that earlier)

@hoegaarden ack - thank you!

**Comment by saad-ali** [comment]
Thanks for working on this @hoegaarden. Ping me on slack when it is ready to review. 

> @lavalamp requested the CRD be defined in a yaml file in the original API PR, and I agree with that request. Reviewing and tracking changes to API type definitions made in code like this is difficult and error prone.
> 
> Additionally, that yaml file should be in a package covered by API review and approval

A separate yaml file would be nice to have, but for the sake of release blocker bug, not a requirement (no user visible impact) -- if it is too much work, we can split the file out in the next release, if needed.

**Review by saad-ali — APPROVED** [review]
/lgtm
/approve

**Pull request closed without merging**